### ISSUE TITLE
Equihash 192 and 144

### ISSUE THREAD

**Issue by robl45** [MultiPoolMiner/MultiPoolMiner#2028]
I downloaded the brand new master and got it working and its showing tons of algorithms but doesn't show Equihash 192.  I added it to the start.bat and its in the algorithm.txt.  What am I doing wrong?

**Comment by robl45** [comment]
Heres what I am getting now.  It appears both the miner and pool files are messed up somehow.  I attempted to add the things it says in 2002 to make them correct.  As you can see, the other algorithms appear to be benchmarking but equihash192 never shows up.

[MultiPoolMiner_2018-07-20_22-48-59.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2215832/MultiPoolMiner_2018-07-20_22-48-59.txt)

**Comment by robl45** [comment]
Seems I got it working a bit, but I'm missing the --Pers part and the server is not adding algorithm before mine.zergpool   

Status: Idle

Last Speed Active                      Launched Miner               Command
---------- ------                      -------- -----               -------
           00 Days 00 Hours 02 Minutes Once     EWBF2-GPU#01-GPU#02 Bin\NVIDIA-EWBF2-Equihash\miner.exe --algo 144_5 --eexit 1 --api 127.0.0.1:42102 --server mine.zergpool.com --port 2146 --user LeTUNYViANaPTypGuUpCKXihwghJyK1nE2
                                                                    --pass rob,c=LTC --fee 0 --intensity 64 --cuda_devices 01
           00 Days 00 Hours 06 Minutes Once     EWBF2-GPU#01-GPU#02 Bin\NVIDIA-EWBF2-Equihash\miner.exe --algo 192_7 --eexit 1 --api 127.0.0.1:42102 --server mine.zergpool.com --port 2144 --user LeTUNYViANaPTypGuUpCKXihwghJyK1nE2
                                                                    --pass rob,c=LTC --fee 0 --intensity 64 --cuda_devices 01

**Comment by robl45** [comment]
looking in the miner file, it appears its doing everything correct except passing the --pers which doesn't seem to be in the command line.

**Comment by robl45** [comment]
I have not modified anything.  I've just been trying to use the files posted in here to make the equihash work.  I"m guessing when there is code and I"m pasting it in, I"m probably doing it wrong.  Are there complete files that I can just put in instead of going to the issues and having to cut and paste ?

**Comment by UselessGuru** [comment]
You are not using original files - look at the log and it clearly states line 24. And what does the file posted here have there? An ampty line. ???

**Comment by robl45** [comment]
These are the two EWBF files I have and neither seems to be doing the--pers which apparently is required for Zergpool.  I fixed the algorithm part of the server name for Zergpool buy changing the host in the file.  Everything appears to be working fine accept the new equihashes.  And to reiterate, I have modified nothing, I'm just trying to mine equihash 192 and the other varients and according to here, we had to add some other files.

[Ewbf.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2216338/Ewbf.txt)

[EWBF2.zip](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2216339/EWBF2.zip)

**Comment by robl45** [comment]
BTW line 24 is the exact same thing in hte file I uploaded here and the file you gave me a few days ago that you said was from your version.  

$Coins = [PSCustomObject]@{   Thats line 24

**Comment by UselessGuru** [comment]
I am sorry, but you are NOT using the file which is in this PR!

https://github.com/MultiPoolMiner/MultiPoolMiner/pull/2002/files#diff-d36e60f6c98fcaf32bca03a9c9c46afaR57

Line 24 is empty

**Comment by robl45** [comment]
Okay, found the one you were talking about.  Attached for reference.  Still is not passing the PERS command.  Logs attached for reference.  It is also passing the wrong algorithm name for Zergpool as Zergpool wants Equihash192 and not Equihash1927.mine.zergpool.  I modified the entries in algorithm.txt and it then passed the correct name as you can see in the logs  but I'm concerned that will mess with other pools potentially.   So the only modification I have made to the program at this point was changing the host line slightly on the zergpool file so it should insert the algorithm before mine.zergpool.com as Zergpool states it should be.  If it doesn't need that, I'm happy to remove.
[MultiPoolMiner_2018-07-21_15-36-48.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2216441/MultiPoolMiner_2018-07-21_15-36-48.txt)
[MultiPoolMiner_2018-07-21_15-38-50.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2216442/MultiPoolMiner_2018-07-21_15-38-50.txt)


[EWBF2.zip](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2216440/EWBF2.zip)

**Comment by m-bosch** [comment]
just as notice, Zergpool has the following notification at the moment:

> For equihash192 Zero mining command line "miner --algo 192_7 --pers ZERO_PoW --server equihash192.mine.zergpool.com --port 2144 --user wallet"
For equihash144 Snowgem mining command line "miner --algo 144_5 --pers sngemPoW --server equihash144.mine.zergpool.com --port 2146 --user wallet"
All above mined with EWBF Equihash miner v0.4

**Comment by UselessGuru** [comment]
@m-bosch 

This is perfectly fine and MPM does handle this just fine.

**Comment by UselessGuru** [comment]
@robl45 

DON'T modify files and then ask for help please. This makes support extremely difficult.

It will only add the --pers command if the pool supplies the coin name (which currently only MiningPoolHubCoins and Zergpool do, No coin name -  no --pers

In order to use the new miner you need a working set of files as I've listed in https://github.com/MultiPoolMiner/MultiPoolMiner/pull/2002 (first post)

>  It is also passing the wrong algorithm name for Zergpool as Zergpool wants Equihash192 and not Equihash1927.mine.zergpool
Equihash1927.mine.zergpool is rubish. The zergpool file posted does NOT add any algo name.
Please post your ZergPool.ps1.

**Comment by UselessGuru** [comment]
BTW: The PR for the Equihash variants in Algorithms.txt has been committed;
https://github.com/MultiPoolMiner/MultiPoolMiner/blob/master/Algorithms.txt

**Comment by robl45** [comment]
I believe I have all those files and as I asked before if I could get a zip of the files I need then I could test them.  At any rate I'm using  zergpool and it isn't paying the pers switch.   I believe I have all the files you mentioned as well or certainly the ones that pertain to zergpool.

**Comment by UselessGuru** [comment]
@robl45 

Please upload all pool files, algorithm.txt and the problematix EWBF2.ps1. I cannot see what's wrong.

**Comment by robl45** [comment]
Okay I'll get them up a little later.  Thanks

**Comment by robl45** [comment]
[Pools.zip](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2216534/Pools.zip)
[EWBF2.zip](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2216535/EWBF2.zip)
[Algorithms.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2216537/Algorithms.txt)

**Comment by paulpoco** [comment]
Trying the 144 and 192 on Zergpool and at least right now it is only using the 2nd GPU while Benchmarking, will have to check what it does on other Algos and report back if other Algos use both my GPUs.  Using all the PRs to do the new Equihash variants.

**Comment by robl45** [comment]
The only Difference between PaulPoco and mine from looking at his EWBF log and mine is his is putting the "ZERO_PoW" after the algo and mine just says the algo.

**Comment by paulpoco** [comment]
@robl45 I never put anything anywhere, that is the code from the PR.

**Comment by robl45** [comment]
PaulPoco,  Maybe you could share a zip of the files you are using.  I think I'm using the correct ones, can't understand why its not throwing in the extra part.

**Comment by robl45** [comment]
redownloaded the master this morning.  added attached EWBF2 file.  Changed start.bat to have my info and changed config.txt to use LTC for Zergpool.  Let it run.  No Equihash 192 or 144.  Went to Algorithm.txt on here and copied Equihash section into algorithm.txt.  Now it shows the equihash variants but same issue, no --Pers after algo.
[EWBF2.zip](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2217061/EWBF2.zip)

**Comment by robl45** [comment]
Current algorithm.txt and log.  multipoolminer doesn't like something in the algorithm.txt.  It appears to be the extra equihash lines, but without those I don't get the new variants of equihash.

[Algorithms.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2217063/Algorithms.txt)
[MultiPoolMiner_2018-07-22_09-57-39.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2217064/MultiPoolMiner_2018-07-22_09-57-39.txt)

**Comment by robl45** [comment]
Also appears that its picking up some coins for zergpool and not others, maybe that is part of the problem?

![image](https://user-images.githubusercontent.com/34346751/43046396-3ca2c7f0-8d96-11e8-8577-a946de6d1694.png)

**Comment by paulpoco** [comment]
@robl45 

I am using everything from July 18 7:30pm Master plus the following PRs:
![image](https://user-images.githubusercontent.com/801414/43048054-2b30399e-8d96-11e8-9b96-d47d6d001db2.png)

And from MinersLegacy/EWBF2.ps1 change line 57 the part with -join '' --> -join ' '
Otherwise you will end up using just one GPU because it will try to do --cuda_devices 012 etc instead of --cuda_devices 0 1 2.

I am using the combination of the new Equihash & CryptoNight new algorithms.txt.

Everthing I am using is the PRs required for the new Equihashes.

Right now I am only using Zerpool but excluding all other pools.  That is why MPH is showing up.

Equihash1927,equihash1445 in my start.bat.

Your algorithms.txt has spacing issues and more than one algo on one line.
![image](https://user-images.githubusercontent.com/801414/43048125-869acb36-8d97-11e8-80e7-276c5d11310a.png)

**Comment by paulpoco** [comment]
@robl45 some of the PRs I am using might be merged already too.

You cannot really take the file from the PR if you are using a different PR that has changed the same file.  You almost have to look it line by line and insert in the right place.  I.e. if 2 different users do a PR for like say MultiPoolMiner.ps1 and the line numbers don't match up.

I am getting the from the Pull Request tab, and started at the bottom and adding the ones I want to use working my way up.

**Comment by robl45** [comment]
Yes, that is what I'm attempting to do too, never had a real issue before this time.  I believe I have the correct files that is why I asked if someone could just provide the actual files.

**Comment by paulpoco** [comment]
> What does it mean in your screen shot when it says 0 includes PR request xxx, xxy, xxz, etc?

0 at the beginning of the file is just so it is near the top of the directory and there is nothing in the txt files, they are just there to remind me what I merged together.

> You are going to the pull request tab, looking at the request and then picking out from all the text what is needed right or is there a separate place that just has the files needed?

Yes I look at the pull requests and modify the files needed.

**Comment by UselessGuru** [comment]
@robl45 / @paulpoco 

> You cannot really take the file from the PR if you are using a different PR that has changed the same file. You almost have to look it line by line and insert in the right place

There is a lot of changes happening with this Equihash mess (look at the discussion with Nemo). Having sub-algos like CryptoNight is going to cause a lot of issues and confusion for everyone.
For now Algorithms.txt in the master zip file (committed in https://github.com/MultiPoolMiner/MultiPoolMiner/pull/2018)
works for Bitcoin Gold (EquihashBTG) on MiningPoolHub. For ZergPool it MAY WORK. Here we have to wait ans see whaht the pool operators will come up with.

**Comment by nimbosa** [comment]
is this related?

2018-07-23 14:04:28 INFO: Unable to determine divisor for BitcoinZ using **Equihash144btcz** algorithm

**Comment by paulpoco** [comment]
@nimbosa ZergPool hasn't fully updated their API for Equihash144btcz yet.

I tried just running it in CMD.exe and it works.
> Bin\NVIDIA-EWBF2-Equihash>miner.exe --algo 144_5 --pers BitcoinZ --server equihash144btcz.mine.zergpool.com --port 2148 --user LTszTqvw5EQJBkNxpvJo4EiTYybMaNJgq3 --pass multipoolminer,c=LTC --pec --fee 0 --intensity 64 --cuda_devices 0 1

**Comment by UselessGuru** [comment]
> 2018-07-23 14:04:28 INFO: Unable to determine divisor for BitcoinZ using Equihash144btcz algorithm

Yes and not.
It is an error by the Pool API which does not send all required data.

**Comment by UselessGuru** [comment]
>  will not show equihash variants for zergpool.

See below. The synthetic algos provided by zergpool do not get mapped 

### Just for your understanding how the algo mapping works:

**The pool stuff:**

The pool exposes algorithm Equihash144 through the API
This is then mapped by Get-Algorithm to Equihash1445 (ZergPool.ps1 line 37)
The coin name is added (ZergPool.ps1 line 42)
The miner object (http://localhost:3999/allpools) will then have an entry like

        "equihash1445":  {
                             "Algorithm":  "Equihash1445",
                             "CoinName":  "SnowGem",
                             "Price":  3.65349E-06,
                             "StablePrice":  3.4701133791415582E-06,
                             "MarginOfError":  0.000734923530797865,
                             "Protocol":  "stratum+tcp",
                             "Host":  "europe.mine.zergpool.com",
                             "Port":  2146,
                             "User":  "aBcDeFgHiJ1234567890",
                             "Pass":  "BlackBox,c=BTC",
                             "Region":  "Europe",
                             "SSL":  false,
                             "Updated":  "\/Date(1531861963113)\/",
                             "Name":  "ZergPool",
                             "Price_Bias":  3.6481199284589306E-06,
                             "Price_Unbias":  3.65349E-06
                         },
From now on MPM will only work with the normalized algo name Equihash1445 (the result from Get-Algorithm)

**The miner stuff:**

The miner file iterates through the available algorithms (as defined in EWBF2.ps1 lines 17-20)
During the second iteration $_.Algorithm will be 'Equihash-144_5'.
The variable $Algorithm is then filled with 144_5
$Algorithm = ($_.Algorithm) -replace "Equihash-" (EWBF2.ps1 line 43)
The algo Equihash-144_5 gets mapped to Equihash1445 (EWBF2.ps1 line 46)
The miner file finds the pool object for algo Equihash1445 and puts together the command line for the miner. This will then be something like

miner.exe --algo 144_5 --pers sngemPoW --eexit 1 --api 127.0.0.1:42100 --server europe.mine.zergpool.com --port 2146 --user aBcDeFgHiJ1234567890 --pass WorkerName,c=BTC --pec --fee 0 --intensity 64 --cuda_devices 0

**Comment by robl45** [comment]
It was showing 192 and 144 the zergpool the other day.

**Comment by robl45** [comment]
And here it is broken out by the specific equihash algorithm and still not using --pers.  It did use pers with multipool miner I noticed.    I got it to do this by using these files  
[Algorithms.txt](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2231561/Algorithms.txt)
[Start1.zip](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2231565/Start1.zip)

Otherwise it doesn't break them out and from what I can see will never use them.



![image](https://user-images.githubusercontent.com/34346751/43257316-7d7b5738-909d-11e8-8f3c-400536b7d4e7.png)

![image](https://user-images.githubusercontent.com/34346751/43257343-8ca92ef6-909d-11e8-95d4-3d117c0abb63.png)

**Comment by robl45** [comment]
And back to normal files.
![image](https://user-images.githubusercontent.com/34346751/43257592-30cc8af0-909e-11e8-944f-bd6872f7d3a7.png)

![image](https://user-images.githubusercontent.com/34346751/43257616-40e84398-909e-11e8-9feb-ffd30a38e314.png)

**Comment by UselessGuru** [comment]
> And back to normal files.

What is normal files? Where do you get the Algorithms.txt from for this configuration? The file in the master is currently broken, that's probalby why you are seeing errors.

To use all the Equihash stuff as well as CryptoNight you need to use a complete Algorithms.txt which is a merge of
https://github.com/MultiPoolMiner/MultiPoolMiner/blob/master/Algorithms.txt and https://github.com/MultiPoolMiner/MultiPoolMiner/pull/2019

For the lastest Algorithms.txt see here:
https://github.com/MultiPoolMiner/MultiPoolMiner/issues/2048

**Comment by robl45** [comment]
Normal is ones I've been using including the algorithm. Txt you gave me to use and start. Bat with no algorithms specified

**Comment by UselessGuru** [comment]
You confuse me now. Just use Algorithms.txt in https://github.com/MultiPoolMiner/MultiPoolMiner/issues/2048 and you should be fine.

**Comment by paulpoco** [comment]
FYI
@UselessGuru Zergpool has changed the equihash144 to equihash144xsg and added equihash144safe.

equihash144safe   Coin name is Safecoin.   --pers Safecoin.

![image](https://user-images.githubusercontent.com/801414/43273524-813f9988-90b1-11e8-9802-950c8ffd569c.png)

**Comment by UselessGuru** [comment]
> Do all equihash144 "--algo 144_5" have the same hash rate no matter which coin?

So far my understanding. And then there are miners which can mine all coins without the extra --pers:
https://github.com/Optiminer/OptiminerEquihash

**Comment by Byrd1083** [comment]
@UselessGuru I have followed your instructions to get 144 and 192 working with Zergpool. I have been able to get it to benchmark both algo's with Zergpool. However, I still does not benchmark or report profitability from Miningpoolhub. MPH reports basic Equihash as expected but does not benchmark 144_5. Any ideas? I am using current master file with no modifications to any scrypts other then start.bat and the upto date algo text file which I have not modified either. I also implemented the EBWF2.ps1 file as provided.
